src/lib.rs

```diff
@@ -1,36 +1,103 @@
 
 use anyhow::{anyhow, Result};
 use guest::prelude::*;
+use k8s_openapi::api::apps::v1::{DaemonSet, Deployment, ReplicaSet, StatefulSet};
+use k8s_openapi::api::batch::v1::CronJob;
+use k8s_openapi::api::batch::v1::Job;
 use k8s_openapi::api::core::v1 as apicore;
+use k8s_openapi::api::core::v1::Pod;
+use k8s_openapi::api::core::v1::ReplicationController;
+use k8s_openapi::apimachinery::pkg::apis::meta::v1::ObjectMeta;
+use k8s_openapi::Resource;
 use kubewarden_policy_sdk::wapc_guest as guest;
+use lazy_static::lazy_static;
 extern crate kubewarden_policy_sdk as kubewarden;
-use kubewarden::{protocol_version_guest, request::ValidationRequest, validate_settings};
+use kubewarden::{
+    protocol_version_guest, request::KubernetesAdmissionRequest, request::ValidationRequest,
+    validate_settings,
+};
 
 mod settings;
 use crate::settings::Rule;
 
-const ALL_ARE_USED_ERROR_MSG: &str = "Rule environment variables are not defined in the resource.";
-const NOT_ALL_ARE_USED_ERROR_MSG: &str = "Rule environment variables are defined in the resource.";
-const ANY_IN_ERROR_MSG: &str = "Resource misses at least one environment variable from the rule.";
-const ANY_NOT_IN_ERROR_MSG: &str =
+static ALL_ARE_USED_ERROR_MSG: &str = "Rule environment variables are not defined in the resource.";
+static NOT_ALL_ARE_USED_ERROR_MSG: &str = "Rule environment variables are defined in the resource.";
+static ANY_IN_ERROR_MSG: &str = "Resource misses at least one environment variable from the rule.";
+static ANY_NOT_IN_ERROR_MSG: &str =
     "Resource should not contain any environment variable from the rule.";
 
+static EXTRACT_LABEL_ANNOTATION_REGEX_STR: &str = r#"\[['"](.*)['"]\]"#;
+lazy_static! {
+    static ref EXTRACT_LABEL_ANNOTATION_REGEX: Regex =
+        Regex::new(EXTRACT_LABEL_ANNOTATION_REGEX_STR).unwrap();
+}
+
 #[no_mangle]
 pub extern "C" fn wapc_init() {
     register_function("validate", validate);
     register_function("validate_settings", validate_settings::<settings::Settings>);
     register_function("protocol_version", protocol_version_guest);
 }
 
-fn validate_envvar_with_rule(rule: &Rule, env_vars: &Vec<apicore::EnvVar>) -> Result<()> {
+fn validate_envvar_with_rule(
+    rule: &Rule,
+    pod: &apicore::PodSpec,
+    metadata: &ObjectMeta,
+    env_vars: &Vec<apicore::EnvVar>,
+) -> Result<()> {
     let error_message: &str;
     let mut resource_env_var: HashSet<settings::EnvVar> = HashSet::new();
     for envvar in env_vars {
-        resource_env_var.insert(settings::EnvVar {
-            name: envvar.name.clone(),
-            value: envvar.value.clone(),
-        });
+        if let Some(env_var_source) = &envvar.value_from {
+            if let Some(field_ref) = &env_var_source.field_ref {
+                let mut envvar_value = None;
+                let field_path = field_ref.field_path.as_str();
+                if field_path == "metadata.name" {
+                    envvar_value = metadata.name.clone();
```

Comment:
The goal of this change is to allow users to have an alternative for [NV features](https://github.com/kubewarden/kubewarden-controller/issues/305). This is was NV admission controller does. It checks for the high level resources as well. If we decide to do something else, I'm fine. But it will diverge from what NV does.